#175636850 Add route for creating a multi city trip request and for a user to get trip requests he made.

hervebutera opened the pull request:
#### What does this PR do?
This PR adds route for creating a trip request and which supports multi city trip.
Also adds route for retrieving trip requests made by a user.
#### Description of Task to be completed?
Adds functionality for accepting multiple trips in one trip request.
Only logged in users can make a trip request.
Also adds functionality for retrieving trip requests made by a user.
#### How should this be manually tested?
- clone the repo by `git clone https://github.com/atlp-rwanda/tech-strikers-bn-backend.git`
- run `cd tech-strikers-bn-backend`
- run `git checkout `ft-multi-city-trip-request-175636850`
- run `npm install`
#### Any background context you want to provide?
The created route has token verification i.e adding a token header.
#### What are the relevant pivotal tracker stories?
[#175636850](https://www.pivotaltracker.com/story/show/175636850) Story for a user to make a multi city trip request. 
[#175636851](https://www.pivotaltracker.com/story/show/175636851) Story for a user being able to retrieve trip requests he made. 
#### Screenshots (if appropriate)
Postman screenshot when successfully posted a trip request
<img width="523" alt="Capture" src="https://user-images.githubusercontent.com/48241489/104951964-40c84800-59cc-11eb-8bd2-fafd120f47b4.PNG">


Postman screenshot when that user has retrieved trip requests he has made.
<img width="540" alt="Capture1" src="https://user-images.githubusercontent.com/48241489/104951936-33ab5900-59cc-11eb-9e83-8ba283676ab1.PNG">


Postman screenshot that shows validation error when `returnDate` is less than `departureDate`
<img width="955" alt="Capture" src="https://user-images.githubusercontent.com/48241489/104329438-921c9700-54f5-11eb-9035-b069da58181a.PNG">


#### Questions:

bigelio commented:
Can you take an action on multiple languages and documenting this apis with swagger?

hervebutera commented:
> Can you take an action on multiple languages and documenting this apis with swagger?

Done!

Duncanwar commented on src/database/migrations/20201220125007-create-location.js:
```diff
@@ -0,0 +1,30 @@
+  down: async (queryInterface, Sequelize) => {
+    await queryInterface.dropTable('Locations');
+  }
+};
```
add new line

Duncanwar commented on src/routes/api/tripRequest.route.js:
```diff
@@ -0,0 +1,10 @@
+const router = express.Router();
+const { sendTripRequest } = tripRequestController;
+
+router.post('/tripRequest/:userToken', tokenVerify, sendTripRequest);
```
why are using user token?

hervebutera replied:
I used the user token to get the `user id` from it. But now corrected!

Duncanwar commented on src/services/trip.service.js:
```diff
@@ -0,0 +1,11 @@
+export default class tripService {
+    static async createTrip (tripData) {
+        let tripInfo;
+        await Trips.create(tripData).then(trip => tripInfo = trip.dataValues)
```
```suggestion
      const trip =  await Trips.create(tripData);
    tripinfo  = await  trip.dataValues;
```

Duncanwar commented on src/services/tripRequest.service.js:
```diff
@@ -0,0 +1,25 @@
+export default class tripRequestService {
+    static async createTripRequest (tripData) {
+        let newTripRequestId;
+        await TripRequest.create(tripData)
```
create variable to avoid using then in async await

Duncanwar commented on src/services/tripRequest.service.js:
```diff
@@ -0,0 +1,25 @@
+
+    static async getTripRequestByUserId (id) {
+        let response;
+        await TripRequest.findOne(
```
same here remove then create variables and use await

Duncanwar commented on src/utils/responses.js:
```diff
@@ -7,7 +7,7 @@
  * @return {object} return json object of successfully http verb action
  */
-const successResponse = (res, code, token, message, data = null) => res.status(code).json({
+const successResponse = (res, code, token, message, data) => res.status(code).json({
```
why have you removed `data` = null

Duncanwar reviewed (changes requested):
Implement those feedback

Duncanwar commented on src/database/migrations/20201215155240-create-trip-request.js:
```diff
@@ -0,0 +1,43 @@
+      },
+      status: {
+        type: Sequelize.STRING,
+        defaultValue: "open"
```
Can we use Pending it sounds logical?

Duncanwar commented on src/controllers/trip.request.js:
```diff
@@ -0,0 +1,44 @@
+
+export default class tripRequestControllers { 
+    static async sendTripRequest (req, res) {
+        if (req.user != undefined) {
```
Make a middleware for checking these conditions

Duncanwar commented on src/database/migrations/20201221203940-create-trip-request.js:
```diff
@@ -0,0 +1,44 @@
+        type: Sequelize.DATE
+      }
+    });
+  },
```
can we add is isApproved field?

hervebutera replied:
> can we add is isApproved field?

That would be a repetition because `status ` field may have values like `pending` as default , `approved or `denied`.

Duncanwar commented on src/middlewares/tokenAuthentication.js:
```diff
@@ -6,8 +6,9 @@ const verifyToken = (req, res, next) => {
-  const user = jwt.decode(token, process.env.ACCESS_TOKEN_SECRET);
+  if (!token) { return res.sendStatus(401) }
+  
+  const user = jwt.verify(token, process.env.ACCESS_TOKEN_SECRET, {expiresIn: "24h"}) 
```
this function was created in util.jwt

hervebu replied:
@Duncanwar  I replaced `jwt.decode` with `jwt.verify()` because `jwt.decode()` doesn't check if the token is expired!

Duncanwar commented on src/validation/userUpdate.validation.js:
```diff
@@ -0,0 +1,10 @@
+    "password": Joi.string().min(6),
+})
+
+export default { updateUserInfoValidation }
```
new line

Duncanwar commented:
you can call the function

On Wed, Jan 6, 2021 at 8:23 PM Herve Butera <notifications@github.com>
wrote:

> *@hervebu* commented on this pull request.
> ------------------------------
>
> In src/middlewares/tokenAuthentication.js
> <https://github.com/atlp-rwanda/tech-strikers-bn-backend/pull/38#discussion_r552886259>
> :
>
> > @@ -6,8 +6,9 @@ const verifyToken = (req, res, next) => {
>      return res.sendStatus(403);
>    }
>    const token = authHeader.split(" ")[1];
> -
> -  const user = jwt.decode(token, process.env.ACCESS_TOKEN_SECRET);
> +  if (!token) { return res.sendStatus(401) }
> +
> +  const user = jwt.verify(token, process.env.ACCESS_TOKEN_SECRET, {expiresIn: "24h"})
>
> @Duncanwar <https://github.com/Duncanwar> I replaced jwt.decode with
> jwt.verify() because jwt.decode() doesn't check if the token is expired!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/atlp-rwanda/tech-strikers-bn-backend/pull/38#discussion_r552886259>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANYV2R2VVL7UOI6PMMS6IXLSYSTDJANCNFSM4U52ZHNQ>
> .
>

hervebu commented:
> you can call the function
> […](#)
> On Wed, Jan 6, 2021 at 8:23 PM Herve Butera ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In src/middlewares/tokenAuthentication.js <[#38 (comment)](https://github.com/atlp-rwanda/tech-strikers-bn-backend/pull/38#discussion_r552886259)> : > @@ -6,8 +6,9 @@ const verifyToken = (req, res, next) => { return res.sendStatus(403); } const token = authHeader.split(" ")[1]; - - const user = jwt.decode(token, process.env.ACCESS_TOKEN_SECRET); + if (!token) { return res.sendStatus(401) } + + const user = jwt.verify(token, process.env.ACCESS_TOKEN_SECRET, {expiresIn: "24h"}) @Duncanwar <https://github.com/Duncanwar> I replaced jwt.decode with jwt.verify() because jwt.decode() doesn't check if the token is expired! — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#38 (comment)](https://github.com/atlp-rwanda/tech-strikers-bn-backend/pull/38#discussion_r552886259)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ANYV2R2VVL7UOI6PMMS6IXLSYSTDJANCNFSM4U52ZHNQ> .

Yeah a good idea!

Duncanwar commented:
If possible can you increase your coveralls?

fabriceumuhire commented on src/controllers/trip.request.js:
```diff
@@ -0,0 +1,42 @@
+                            await createTrip(tripDetails);
+                        } 
+                    }
+                    return successResponse(res, created,null, res.__("tripRequestSubmitted"))
```
Great work pal, the PR looks good for me.
Can we have the response message only? I believe we do not need to have a response with a token while creating a trip request.

fabriceumuhire reviewed (commented):
The PR looks good, just wondering about the response while creating a trip request which includes the token.

hervebutera commented:
> The PR looks good, just wondering about the response while creating a trip request which includes the token.

Solved it, now there is a response function which doesn't have the token!

hervebutera commented:
> If possible can you increase your coveralls?

Tried the best as of now!

Duncanwar commented:
Resolve conflicts

hervebutera committed:
feat(tripRequest): multicity trip request route and get trip requests route by id.

add route for making a multi city trip request
add route that retrieve trip requests by user id.

[delivers #175636850]
